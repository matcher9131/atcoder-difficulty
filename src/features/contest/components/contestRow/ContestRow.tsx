import type { ReactElement, ReactNode } from "react";

export type ContestRowProps = {
    readonly headerCell: ReactElement;
    readonly problemCells: readonly ReactElement[];
};

export const ContestRow = ({ headerCell, problemCells }: ContestRowProps): ReactNode => {
    return (
        <tr className="contents">
            {headerCell}
            {problemCells}
        </tr>
    );
};
