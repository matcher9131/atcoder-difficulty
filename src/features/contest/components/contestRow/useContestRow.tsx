import { useAtom } from "jotai";
import type { ContestRowProps } from "./ContestRow";
import { ContestHeaderCellContainer } from "../contestHeaderCell";
import { contestProblemIdsAtom } from "../../../problem/dict/problems";
import { ProblemCellContainer } from "../../../problem/components/problemCell";

export const useContestRow = (
    contestId: string,
    numProblems: number,
): ContestRowProps => {
    const headerCell = <ContestHeaderCellContainer contestId={contestId} />;
    const [problemIds] = useAtom(contestProblemIdsAtom(contestId));
    const problemCells = problemIds.map((problemId) => (
        <ProblemCellContainer key={problemId} problemId={problemId} />
    ));
    return { numProblems, headerCell, problemCells };
};
