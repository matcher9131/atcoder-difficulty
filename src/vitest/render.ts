import { render as originalRender, type RenderOptions } from "@testing-library/react";
import userEvent, { type UserEvent } from "@testing-library/user-event";
import type { ReactNode } from "react";

export const render = (
    element: ReactNode,
    options?: Omit<RenderOptions, "queries">,
): ReturnType<typeof originalRender> & { readonly user: UserEvent } => {
    return {
        user: userEvent.setup(),
        ...originalRender(element, options),
    };
};
