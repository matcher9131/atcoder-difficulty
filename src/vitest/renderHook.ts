import { renderHook as originalRenderHook } from "@testing-library/react";
import { Provider } from "jotai";

export const renderHook = <T, Props>(
    render: (initialProps: Props) => T,
): ReturnType<typeof originalRenderHook<T, Props>> => {
    return originalRenderHook(render, { wrapper: Provider });
};
