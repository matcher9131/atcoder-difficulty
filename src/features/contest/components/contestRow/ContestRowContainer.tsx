import type { ReactNode } from "react";
import { ContestRow } from "./ContestRow";
import { useContestRow } from "./useContestRow";

type ContestRowContainerProps = {
    readonly contestId: string;
    readonly numProblems: number;
};

export const ContestRowContainer = ({
    contestId,
    numProblems,
}: ContestRowContainerProps): ReactNode => {
    const props = useContestRow(contestId, numProblems);
    return <ContestRow {...props} />;
};
