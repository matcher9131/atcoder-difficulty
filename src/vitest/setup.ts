import * as matchers from "@testing-library/jest-dom/matchers";
import type { TestingLibraryMatchers } from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { afterEach, expect } from "vitest";

// Extend Vitest's expect with jest-dom matchers
expect.extend(matchers);

// Merge type declaration of matchers of jest-dom to vitest
declare module "vitest" {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    interface Assertion<T = any> extends jest.Matchers<void, T>, TestingLibraryMatchers<T, void> {}
}

// Clean up the DOM after each test
afterEach(() => {
    cleanup();
});
