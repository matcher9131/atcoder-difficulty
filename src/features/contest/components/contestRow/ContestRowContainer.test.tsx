import { describe, it, expect, vi, beforeEach } from "vitest";
import { render } from "../../../../vitest/render";
import { ContestRowContainer } from "./ContestRowContainer";
import { useContestRow } from "./useContestRow";
import { ContestRow } from "./ContestRow";

// Mock useContestRow and ContestRow
const mockRowProps = {
    headerCell: <td data-testid="header-cell">Header</td>,
    problemCells: [
        <td key="1" data-testid="problem-cell-1">
            Problem 1
        </td>,
        <td key="2" data-testid="problem-cell-2">
            Problem 2
        </td>,
    ],
};

vi.mock("./useContestRow", () => ({
    useContestRow: vi.fn(),
}));

vi.mock("./ContestRow", () => ({
    ContestRow: vi.fn(() => <tr data-testid="contest-row" />),
}));

describe("ContestRowContainer", () => {
    const contestId = "abc123";
    const numProblems = 2;

    beforeEach(() => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestRow as any).mockClear();
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (ContestRow as any).mockClear();
    });

    it("calls useContestRow with contestId and numProblems", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestRow as any).mockReturnValue(mockRowProps);
        render(<ContestRowContainer contestId={contestId} numProblems={numProblems} />);
        expect(useContestRow).toHaveBeenCalledWith(contestId, numProblems);
    });

    it("renders ContestRow with props from hook", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestRow as any).mockReturnValue(mockRowProps);
        render(<ContestRowContainer contestId={contestId} numProblems={numProblems} />);
        expect(ContestRow).toHaveBeenCalledWith(mockRowProps, undefined);
    });

    it("matches snapshot", () => {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-call, @typescript-eslint/no-unsafe-member-access, @typescript-eslint/no-explicit-any
        (useContestRow as any).mockReturnValue(mockRowProps);
        const { container } = render(<ContestRowContainer contestId={contestId} numProblems={numProblems} />);
        expect(container).toMatchSnapshot();
    });
});
