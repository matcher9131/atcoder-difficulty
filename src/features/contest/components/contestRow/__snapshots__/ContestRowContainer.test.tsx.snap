// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ContestRowContainer > matches snapshot 1`] = `
<div>
  <tr
    data-testid="contest-row"
  />
</div>
`;
